src/workerd/io/container-client.c++

```diff
@@ -0,0 +1,334 @@
+    kj::Own<kj::HttpClient> httpClient,
+    kj::String containerName,
+    kj::String imageName)
+    : byteStreamFactory(byteStreamFactory),
+      timer(timer),
+      network(kj::mv(network)),
+      httpClient(kj::mv(httpClient)),
+      containerName(kj::mv(containerName)),
+      imageName(kj::mv(imageName)) {}
+
+// Docker-specific Port implementation that implements rpc::Container::Port::Server
+class ContainerClient::DockerPort final: public rpc::Container::Port::Server {
+ public:
+  DockerPort(ContainerClient& containerClient, kj::String containerHost, uint16_t containerPort)
+      : containerClient(containerClient),
+        containerHost(kj::mv(containerHost)),
+        containerPort(containerPort) {}
+
+  kj::Promise<void> connect(ConnectContext context) override {
+    kj::HttpHeaderTable headerTable;
+    kj::HttpHeaders headers(headerTable);
+
+    // Port mappings might have outdated mappings, we can't know if a connect request
+    // fails because the app hasn't finished starting up or because the mapping is outdated.
+    // To be safe we should inspect the container to get up to date mappings.
+    auto portMappings = kj::get<1>(co_await containerClient.inspectContainer());
+    auto maybeMappedPort = portMappings.find(containerPort);
+    if (maybeMappedPort == kj::none) {
+      throw KJ_EXCEPTION(DISCONNECTED, "connect(): Connection refused: container port not found");
+    }
+    auto mappedPort = KJ_ASSERT_NONNULL(maybeMappedPort);
+
+    auto request = containerClient.network->connect(
+        kj::str(containerHost, ":", mappedPort), headers, kj::HttpConnectSettings());
+
+    auto status = co_await request.status;
+    if (status.statusCode != 200) {
+      throw KJ_EXCEPTION(DISCONNECTED, "connect(): Connection refused");
+    }
+
+    auto upPipe = kj::newOneWayPipe();
+    auto upEnd = kj::mv(upPipe.in);
+    auto results = context.getResults();
+    results.setUp(containerClient.byteStreamFactory.kjToCapnp(kj::mv(upPipe.out)));
+    auto downEnd = containerClient.byteStreamFactory.capnpToKj(context.getParams().getDown());
+    pumpTask = kj::joinPromisesFailFast(
+        kj::arr(upEnd->pumpTo(*request.connection), request.connection->pumpTo(*downEnd)))
+                   .ignoreResult()
+                   .attach(kj::mv(upEnd), kj::mv(request.connection), kj::mv(downEnd));
+    co_return;
+  }
+
+ private:
+  ContainerClient& containerClient;
+  kj::String containerHost;
+  uint16_t containerPort;
+  kj::Maybe<kj::Promise<void>> pumpTask;
+};
+
+kj::Promise<ContainerClient::Response> ContainerClient::dockerApiRequest(
+    kj::HttpMethod method, kj::StringPtr endpoint, kj::Maybe<kj::StringPtr> body) {
+  kj::HttpHeaderTable::Builder builder;
+  const auto userAgent = builder.add("User-Agent");
+  const auto accept = builder.add("Accept");
+  const auto contentType = builder.add("Content-Type");
+  const auto contentLength = builder.add("Content-Length");
+  auto headerTable = builder.build();
```

Comment:
Probably better to build this once and reuse it rather than building on each call